applications/asset_tracker_v2/src/modules/location_module.c

```diff
@@ -0,0 +1,651 @@
+
+		location_module_event->data.agps_request = event_data->agps_request;
+		location_module_event->type = LOCATION_MODULE_EVT_AGPS_NEEDED;
+		APP_EVENT_SUBMIT(location_module_event);
+		break;
+
+	case LOCATION_EVT_GNSS_PREDICTION_REQUEST:
+		LOG_DBG("Getting location assistance requested (P-GPS). Not doing anything.");
+		break;
+
+	case LOCATION_EVT_NEIGHBOR_CELL_REQUEST:
+		LOG_DBG("Getting cellular request");
+		send_neighbor_cell_update((
+			struct lte_lc_cells_info *)&event_data->cellular_request);
+		break;
+
+	default:
+		LOG_DBG("Getting location: Unknown event");
+		break;
+	}
+}
+
+static bool location_data_requested(enum app_module_data_type *data_list, size_t count)
+{
+	for (size_t i = 0; i < count; i++) {
+		if (data_list[i] == APP_DATA_LOCATION) {
+			return true;
+		}
+	}
+
+	return false;
+}
+
+static int setup(void)
+{
+	int err;
+
+	err = location_init(location_event_handler);
+	if (err) {
+		printk("Initializing the Location library failed, error: %d\n", err);
+		return -1;
+	}
+
+	return 0;
+}
+
+/* Message handler for STATE_INIT. */
+static void on_state_init(struct location_msg_data *msg)
+{
+	if (IS_EVENT(msg, modem, MODEM_EVT_INITIALIZED)) {
+		int err;
+
+		err = setup();
+		if (err) {
+			LOG_ERR("setup, error: %d", err);
+			SEND_ERROR(location, LOCATION_MODULE_EVT_ERROR_CODE, err);
+		}
+
+		state_set(STATE_RUNNING);
+	}
+}
+
+/* Message handler for SUB_STATE_SEARCH. */
+static void on_state_running_location_search(struct location_msg_data *msg)
+{
+	if (IS_EVENT(msg, location, LOCATION_MODULE_EVT_INACTIVE)) {
+		sub_state_set(SUB_STATE_IDLE);
+	}
+
+	if (IS_EVENT(msg, app, APP_EVT_DATA_GET)) {
+		if (!location_data_requested(msg->module.app.data_list, msg->module.app.count)) {
+			return;
+		}
+
+		LOG_WRN("Location request already active and will not be restarted");
+		LOG_WRN("Try setting a sample/publication interval greater");
+		LOG_WRN("than the location search timeout.");
+	}
+}
+
+/* Message handler for SUB_STATE_IDLE. */
+static void on_state_running_location_idle(struct location_msg_data *msg)
+{
+	if (IS_EVENT(msg, location, LOCATION_MODULE_EVT_ACTIVE)) {
+		sub_state_set(SUB_STATE_SEARCH);
+	}
+
+	if (IS_EVENT(msg, app, APP_EVT_DATA_GET)) {
+		if (!location_data_requested(msg->module.app.data_list, msg->module.app.count)) {
+			return;
+		}
+
+		search_start();
+	}
+}
+
+/* Message handler for all states. */
+static void on_all_states(struct location_msg_data *msg)
+{
+	if (IS_EVENT(msg, app, APP_EVT_START)) {
+		int err;
+
+		err = module_start(&self);
+		if (err) {
+			LOG_ERR("Failed starting module, error: %d", err);
+			SEND_ERROR(location, LOCATION_MODULE_EVT_ERROR_CODE, err);
+		}
+
+		state_set(STATE_INIT);
+	}
+
+	if (IS_EVENT(msg, util, UTIL_EVT_SHUTDOWN_REQUEST)) {
+		/* The module doesn't have anything to shut down and can
+		 * report back immediately.
+		 */
+		SEND_SHUTDOWN_ACK(location, LOCATION_MODULE_EVT_SHUTDOWN_READY, self.id);
+		state_set(STATE_SHUTDOWN);
```

Comment:
I'm looking at what are actions other modules take in showdown event:
- `modem_module`: `lte_lc_power_off()`
- `ui_module`: some LED state changes which is understandable so that user is shown correct LED indicators

Based on this and as there is reboot coming soon after this, `location_cancel()` shouldn't be necessary. Whether it's useful for some reason is another question. I'd compare potentially active GNSS to cloud connection, which is not disconnected in shutdown event of `cloud_module`. Earlier gnss_module didn't stop GNSS.